compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -4104,9 +4115,26 @@ public void visit(BLangInferredTypedescDefaultNode inferTypedescExpr) {
         node.accept(this);
         parent = myParent;
         checkAccess(node);
+        checkExpressionValidity(node);
         this.env = prevEnv;
     }
 
+    private  <E extends BLangExpression> void checkExpressionValidity(E exprNode) {
+        if (!types.isNeverTypeOrStructureTypeWithARequiredNeverMember(exprNode.getBType())) {
+            return;
+        }
+        if (!checkExpressionInValidParent(exprNode.parent)) {
+            dlog.error(exprNode.pos, DiagnosticErrorCode.EXPRESSION_OF_NEVER_TYPE_NOT_ALLOWED);
+        }
+    }
+
+    private boolean checkExpressionInValidParent(BLangNode currentParent) {
+        return currentParent != null && (currentParent.getKind() == NodeKind.EXPRESSION_STATEMENT ||
+                (currentParent.getKind() == NodeKind.VARIABLE &&
+                        ((BLangSimpleVariable) parent).typeNode.getBType().tag == TypeTags.FUTURE)
+                || currentParent.getKind() == NodeKind.TRAP_EXPR);
```

Comment:
We need to consider the possibility of the expression being within parentheses also.

The following doesn't work with these changes.

```ballerina
public function main() {
    error? x = trap (fn());
}

function fn() returns never {
    panic error("");
}
```